File: python/DUELink/Digital.py

from DUELink.SerialInterface import SerialInterface

class DigitalController:    

    def __init__(self, serialPort:SerialInterface):
        self.serialPort = serialPort

    def Read(self, pin: int, inputType: 0) -> bool:
        if pin < 0 or (pin >= self.serialPort.DeviceConfig.MaxPinIO and pin != 97 and pin != 98 and pin != 108): #A, B, Led
            raise ValueError("Invalid pin")

        pull = "0"
        if inputType == 1:
            pull = "1"
        elif inputType == 2:
            pull = "2"

        cmd = f"print(dread({pin},{pull}))"
        self.serialPort.WriteCommand(cmd)

        respone = self.serialPort.ReadRespone()

        if respone.success:            
            try:
                value = int(respone.respone)
                return value == 1
            except:
                pass

        return False

    def Write(self, pin: int, value: bool) -> bool:
        if pin < 0 or (pin >= self.serialPort.DeviceConfig.MaxPinIO and pin != 108): # Led
            raise ValueError("Invalid pin")

        cmd = f"dwrite({pin},{1 if value else 0})"
        self.serialPort.WriteCommand(cmd)

        respone = self.serialPort.ReadRespone()

        return respone.success

File: dotnet/GHIElectronics.DUELink/Digital.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using static GHIElectronics.DUELink.DUELinkController.DigitalController;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {
        public enum InputType {
            None = 0,
            PullUp = 1,
            PullDown = 2,
        }
        public class DigitalController {

           

            SerialInterface serialPort;

            public DigitalController(SerialInterface serialPort) => this.serialPort = serialPort;

          

            public bool Read(int pin, InputType inputType = InputType.None) {
                if (pin < 0 || (pin >= this.serialPort.DeviceConfig.MaxPinIO ))
                    throw new ArgumentOutOfRangeException("Invalid pin.");

                var input = (int)inputType;

              
            

                var cmd = string.Format("dread({0},{1})", pin.ToString(), input.ToString());



                this.serialPort.WriteCommand(cmd);

                var response = this.serialPort.ReadResponse();

                if (response.success) {

                   
                    try {
                        var value = int.Parse(response.response);

                        return value == 1;
                    }
                    catch { }

                    
                }

                return false;
            }
            

            public bool Write(int pin, bool value) {
                if (pin < 0 || (pin >= this.serialPort.DeviceConfig.MaxPinIO ))
                    throw new ArgumentOutOfRangeException("Invalid pin.");

                var v = (value == true ? 1 : 0);

                var cmd = string.Format("dwrite({0},{1})", pin, (value == true ? 1 : 0));

                this.serialPort.WriteCommand(cmd);

                var response = this.serialPort.ReadResponse();

                return response.success;
            }

           
        }

       
    }
}
